File: leetcode/Dynamic Programming/longest_common_subsequence.c
// https://leetcode.com/problems/longest-common-subsequence/

#define MAX(a,b) ((a) > (b)) ? a : b

int longestCommonSubsequence(char* text1, char* text2) {
    // used this video https://youtu.be/Ua0GhsJSlWM?si=FhBhPQbNKOw4Iu1Z to solve the problem

    int dp[strlen(text1)+1][strlen(text2)+1], row, col;
    memset(dp, 0, sizeof(dp));

    for (row = strlen(text1) - 1; row >= 0; row--) {
        for (col = strlen(text2) - 1; col >= 0; col--) {
            if (text1[row] == text2[col]) {
                dp[row][col] = 1 + dp[row+1][col+1];
            }
            else {
                dp[row][col] = MAX(dp[row+1][col], dp[row][col+1]);
            }
        }
    }
    return dp[0][0];
}

File: leetcode/Dynamic Programming/longest_common_subsequence.py
# https://leetcode.com/problems/longest-common-subsequence/

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [0] * len(text2)
        for char in text1:
            max_pre = 0
            for i in range(len(text2)):
                if text2[i] == char and dp[i] < max_pre + 1:
                    dp[i] = max_pre + 1
                else:
                    max_pre = max(max_pre, dp[i])
        return max(dp)
